### DISCUSSION TITLE
Exchange ideas about the simulation approach

### DISCUSSION THREAD

**Discussion by maximiliankir** [openfootmanager/openfootmanager#10] [Ideas]
The most important part of a football manager game is the simulation engine, as already discussed in #3. I start this discussion to further exchange some ideas about how to compute the outcome of a game and the events, that happen in during the game.

@sturdy-robot brought up some ideas for possible approaches in #3:

- Randomly generate Events from a EventList, based on some probabilities
- Markov Chains
- Agent based simulations
- 2D/3D simulation

My personal opinion is also to start with a simple approach, which can be evolved in the future. But I think that generating random Events from a List of events with a fixed probability is to simple, to get a somehow realistic simulation of a football game.

**Comment by maximiliankir** [comment]
There is a different possible approach, that has not been mentioned yet. I would call it a top-down approach. First calculate the end result of the game, depending on probabilities (depending on the selected fomations, players, fitness ...). Then step by step calculate more fine-granular events. How and when happend the goals, what move was made before the goal ...

But I think this approach is also very limiting. Because you can not influence the outcome of the game during the game, because the result is already calculated.

So my personal opinion is, that we should start with a simple markov based approach. You have a list of possible events and you have a current state. A event leads from the current state to a different state and the probability of a event depends on the current state.
It is open to discussion, what states and events we need. The probabilities for the events should later be calculated dynamically depending on the used player, the selected tactics ...

**Reply by maximiliankir** [reply]
Thank you for your opinion on those ideas. The top-down approach was just one thing I was thinking about. I just wanted to leave it here, as another totaly different approach. But I also don't think that this is a direction in which we should go. At least not at the moment. But maybe at some point we'll start to think differently.

I would model all the different statuses, you mentioned in one state class. Each change in one of the state variables is a new state. For ball location I would divide the pitch in different zones e.g. like this: 
![grafik](https://user-images.githubusercontent.com/44817899/188879990-64f0ba6b-0ea7-46a3-8420-b9334fa3b211.png)

This allows us to still keep a finite number of possible states. And then we would have a huge state transition matrix, which contains all the probabilities for state changes. To stay in the current state is of course also a possible transition.
This transition matrix must be calculated depending on the selected formation and different user choices. After the user paused and changed some settings, this matrix must be recalculated. 
To get started and to test how to tune the probabilities for realistic results, we can start with a static predefined transition matrix. 

There are some more variables, that I would add to the state, e.g.  the current score. And if I think about this more then, I'll come up with more necessary variables for sure. So we should be careful that this is expendable in all directions.

I'm not sure yet, how I would cover special events, like fouls, freekick, corner, penalty, etc. with this state-transition based implementation. But maybe you have a good idea for that.

**Reply by maximiliankir** [reply]
I'll try to make a proof of concept implementation, and I will let you know when I have finished it.

**Reply by maximiliankir** [reply]
I have created a fork of the project and there a branch markov-based, where I started to implement this idea. So if you want to check out how development is going, feel free to have a look. But until now there is not much to see. I just created the basic classes for state and action. And adopted the MatchSimulation and OFM class for that.

I changed my approach slightly, as you can see in the code: I will not go from one state directly to the next state with some probability. But I draw a random action from the current state. And draw then a random outcome of this action (next state). This allows to keep only ball location, team status etc. in the state and implement a duel between to players in an action.

**Reply by maximiliankir** [reply]
I don't know what you are planning to implement this weekend. But for me it would be nice to have the functions that load the teams and players (with their skill values) from the database files. 
That would make it easier for me to test the actions, if I can load some real teams and test the match simulation with that.

[a comment by an automated account is not shown]

**Comment by maximiliankir** [comment]
And in addition we should think about the implementation. I do not think, that it is a good idea to hardcode all the states and events. At the moment the event list and probabilities are hard coded. We should use a generator class, which generates the state machine with the probabilities. And maybe we should also use some kind of description files, which describe the states and the possible events. This keeps the software expendable and improves the code readability.

[a comment by an automated account is not shown]

**Comment by maximiliankir** [comment]
Nice to hear from you and great that you're making progress with the project.
I'm still interested in collaborating, because I think this is an intersting project and has some potential. Unfortunately I don't have much time at the moment and probably will not have much time in the near future.
But we can of course use this discussion thread to discuss design choices. 
To then setup issues, is a good idea in any case. Maybe I can work from time to time on one or someone else discovers this project and helps.

[a comment by an automated account is not shown]

**Comment by tonyemery** [comment]
Hello, it's starting to take on a good shape. Having said that, how are you going to manage the aspects linked to the success of the actions by the players. For example, if 2 players are fighting over the ball, how do you calculate which of the 2 will win? The same goes for a defender tackling an attacker or an attacker shooting at goal.

[a comment by an automated account is not shown]
